#CMAPLE

[![License (GPL v2.0)](https://img.shields.io/badge/GPL%20v2.0-License?logo=gnu&label=License)](https://opensource.org/license/gpl-2-0)


## What's CMAPLE?


CMAPLE is a C++ reimplementation of [MAPLE][MAPLE_FORMAT] - a novel likelihood-based phylogenetic inference method for pandemic-scale epidemiological genomic data. CMAPLE is highly optimized for performance and scalability with many new features.

Apart from a standalone software, we also provide a set of APIs, which allow users to integrate CMAPLE into existing phylogenetic inference methods. For more detail, please see <TBU>**API_Documentation**.


## How to cite CMAPLE?

<TBU>To be updated...


## How to use CMAPLE?


Please refer to our [Wiki page](https://github.com/iqtree/cmaple/wiki) for installation instructions and usage examples of CMAPLE.  


## Contributing - Get Involved!

See [CONTRIBUTING](CONTRIBUTING) file.

## Code of Conduct

CMaple is governed by some rules, which serve to protect our community and its individuals. See [CODE_OF_CONDUCT](CODE_OF_CONDUCT) for details.


## Acknowledgment
CMAPLE relies on the following third-party libraries:

- **[ncl](https://github.com/mtholder/ncl)** (version 2.0) - Licensed under the [GNU General Public License](https://github.com/mtholder/ncl/blob/master/COPYING).
- **[simde](https://github.com/simd-everywhere/simde)** - Licensed under the [MIT License](https://github.com/simd-everywhere/simde/blob/master/COPYING).
- **[zlib](http://zlib.net/)** (version 1.2.7).



[CMAPLE_RELEASES]: https://github.com/trongnhanuit/cmaple/releases
[Gaston et al. 2011]: https://doi.org/10.1093/bioinformatics/btr470
[MAPLE_FORMAT]: https://www.nature.com/articles/s41588-023-01368-0
[DNA_MODELS]: #dna-models
[AA_MODELS]: #amino-acid-models
[TREE_SEARCH]: #4-set-the-tree-search-type
[Guindon et al., 2010]: https://academic.oup.com/sysbio/article/59/3/307/1702850
[IQ_TREE]: http://www.iqtree.org/
[Abascal et al., 2007]: https://doi.org/10.1093/molbev/msl136
[Adachi and Hasegawa, 1996]: https://doi.org/10.1007/BF02498640
[Adachi et al., 2000]: https://doi.org/10.1007/s002399910038
[Bielawski and Gold, 2002]: https://doi.org/10.1093/genetics/161.4.1589
[Dang et al., 2010]: https://doi.org/10.1186/1471-2148-10-99
[Dang et al., 2022]: https://doi.org/10.1093/sysbio/syac007
[Dayhoff et al., 1978]: http://compbio.berkeley.edu/class/c246/Reading/dayhoff-1978-apss.pdf
[Dimmic et al., 2002]: https://doi.org/10.1007/s00239-001-2304-y
[El-Gebali et al., 2018]: https://doi.org/10.1093/nar/gky995
[Felsenstein, 1981]: https://doi.org/10.1007%2FBF01734359
[Goldman and Yang, 1994]: http://mbe.oxfordjournals.org/content/11/5/725.abstract
[Gu et al., 1995]: http://mbe.oxfordjournals.org/content/12/4/546.abstract
[Hasegawa, Kishino and Yano, 1985]: https://dx.doi.org/10.1007%2FBF02101694
[Henikoff and Henikoff, 1992]: https://dx.doi.org/10.1073%2Fpnas.89.22.10915
[Jarvis et al., 2015]: https://doi.org/10.1186/s13742-014-0038-1
[Jones et al., 1992]: https://dx.doi.org/10.1093%2Fbioinformatics%2F8.3.275
[Jukes and Cantor, 1969]: http://doi.org/10.1016/B978-1-4832-3211-9.50009-7
[Kimura, 1980]: https://doi.org/10.1007%2FBF01731581
[Kimura, 1981]: https://doi.org/10.1073/pnas.78.1.454
[Kosiol and Goldman, 2005]: https://doi.org/10.1093/molbev/msi005
[Kosiol et al., 2007]: https://doi.org/10.1093/molbev/msm064
[Lartillot and Philippe, 2004]: https://doi.org/10.1093/molbev/msh112
[Le and Gascuel, 2008]: https://doi.org/10.1093/molbev/msn067
[Le and Vinh, 2020]: https://doi.org/10.1007/s00239-020-09943-3
[Le et al., 2008a]: https://doi.org/10.1093/bioinformatics/btn445
[Le et al., 2008b]: https://doi.org/10.1098/rstb.2008.0180
[Le and Gascuel, 2010]: https://doi.org/10.1093/sysbio/syq002
[Le et al., 2012]: https://doi.org/10.1093/molbev/mss112
[Lewis, 2001]: https://doi.org/10.1080/106351501753462876
[Minh et al., 2021]: https://doi.org/10.1093/sysbio/syab010
[Misof et al., 2014]: https://doi.org/10.1126/science.1257570
[Mueller and Vingron, 2000]: https://doi.org/10.1089/10665270050514918
[Muse and Gaut, 1994]: http://mbe.oxfordjournals.org/content/11/5/715.abstract
[Nickle et al., 2007]: https://dx.doi.org/10.1371/journal.pone.0000503
[Ran et al., 2018]: https://doi.org/10.1098/rspb.2018.1012
[Rota-Stabelli et al., 2009]: https://doi.org/10.1016/j.ympev.2009.01.011
[Schneider et al., 2005]: https://doi.org/10.1186/1471-2105-6-134
[Shen et al., 2018]: https://doi.org/10.1016/j.cell.2018.10.023
[Soubrier et al., 2012]: https://doi.org/10.1093/molbev/mss140
[Tamura and Nei, 1993]: http://mbe.oxfordjournals.org/cgi/content/abstract/10/3/512
[Tavare, 1986]: http://www.damtp.cam.ac.uk/user/st321/CV_&_Publications_files/STpapers-pdf/T86.pdf
[Veerassamy et al., 2004]: https://doi.org/10.1089/106652703322756195
[Vinh et al., 2017]: https://doi.org/10.1186/s12862-017-0987-y
[Wang et al., 2008]: https://doi.org/10.1186/1471-2148-8-331
[Whelan and Goldman, 2001]: https://doi.org/10.1093/oxfordjournals.molbev.a003851
[Woodhams et al., 2015]: https://doi.org/10.1093/sysbio/syv021
[Wu et al., 2018]: https://doi.org/10.1016%2Fj.dib.2018.04.094
[Yang, 1994]: https://doi.org/10.1007/BF00160154
[Yang, 1995]: http://www.genetics.org/content/139/2/993.abstract
[Yang et al., 1998]: http://mbe.oxfordjournals.org/content/15/12/1600.abstract
[Zharkikh, 1994]: https://doi.org/10.1007/BF00160155
[Guindon et al., 2010]: https://doi.org/10.1093/sysbio/syq010